bip-0352/reference.py

```diff
@@ -0,0 +1,219 @@
+#!/usr/bin/env python3
+
+import hashlib
+import json
+import bip32
+from typing import List, Tuple, Dict, Union
+from sys import argv
+
+# local files
+from bech32m import convertbits, bech32_encode, decode, Encoding
+from secp256k1 import ECKey, ECPubKey
+
+
+def sha256(s: Union[bytes, bytearray]) -> bytes:
+    return hashlib.sha256(s).digest()
+
+
+def ser_uint32(u: int) -> bytes:
+    return u.to_bytes(4, 'big')
+
+
+def hash_outpoints(outpoints: List[Tuple[str, int]]) -> bytes:
+
+    outpoints_sorted = sorted([
+        bytearray.fromhex(txid)[::-1] + n.to_bytes(4, 'little')
+        for txid, n in outpoints
+    ])
+
+    s = hashlib.new('sha256')
+    for op in outpoints_sorted:
+        s.update(op)
+
+    return s.digest()
+
+
+def derive_silent_payment_key_pair(seed: bytes) -> Tuple[ECKey, ECKey, ECPubKey, ECPubKey]:
+
+    SCAN_KEY="m/352h/0h/0h/1h/0"
+    SPEND_KEY="m/352h/0h/0h/0h/0"
+
+    master = bip32.BIP32.from_seed(seed)
+    scan = ECKey().set(master.get_privkey_from_path(SCAN_KEY))
+    spend = ECKey().set(master.get_privkey_from_path(SPEND_KEY))
+    Scan = scan.get_pubkey()
+    Spend = spend.get_pubkey()
+
+    return scan, spend, Scan, Spend
+
+
+def encode_silent_payment_address(B_scan: ECPubKey, B_m: ECPubKey, hrp: str = "tsp", version: int = 0) -> str:
+
+    data = convertbits(B_scan.get_bytes(False) + B_m.get_bytes(False), 8, 5)
+    return bech32_encode(hrp, [version] + data, Encoding.BECH32M)
+
+
+def create_labeled_silent_payment_address(B_scan: ECPubKey, B_spend: ECPubKey, m: bytes, hrp: str = "tsp", version: int = 0) -> str:
+
+    G = ECKey().set(1).get_pubkey()
+    B_m = B_spend + m * G
```

Comment:
This seems to not match the description in the BIP, where also the private scan key b_scan and hashing is involved to derive B_m?

Reply:
Updated